### PR TITLE
Checksums work for #3735

### PR DESCRIPTION
**Pull request by ckamm**
Store the checksums in the db during up and downloads.

### CODE REVIEW

File: src/libsync/capabilities.cpp [hunk 1]
```diff
@@ -63,9 +65,28 @@ bool Capabilities::shareResharing() const
...
+QByteArray Capabilities::preferredChecksumType() const
+{
+    auto list = supportedChecksumTypes();
+    if (list.isEmpty()) {
+        return QByteArray();
+    }
+    return list.first();
 }
 
 }
```
Comment (dragotin) [inline src/libsync/capabilities.cpp]: I assume the function `supportedChecksumTypesRaw()` is supposed to fetch the supported checksums from the server capabilities, but that is not yet implemented, right? Why not name it like that? 'raw' is pretty generic imo.

Also, if we have a list of supported ones from the server, the prepend can lead to duplicate entries, right?

File: src/libsync/propagateupload.cpp [hunk 2]
```diff
@@ -209,30 +209,45 @@ void PropagateUploadFileQNAM::start()
...
-    if (validator->checksumType().isEmpty()) {
-        QStringList checksumTypes = _propagator->account()->capabilities().supportedChecksumTypes();
-        if (!checksumTypes.isEmpty()) {
-            // TODO: We might want to prefer some types over others instead
-            // of choosing the first.
-            validator->setChecksumType(checksumTypes.first());
+    auto supportedChecksumTypes = _propagator->account()->capabilities().supportedChecksumTypes();
+
+    // If we already have a checksum header and the checksum type is supported
+    // by the server, we keep that - otherwise recompute.
```
Comment (dragotin) [inline src/libsync/propagateupload.cpp]: minor: maybe add a comment where we get could possibly have the checksum header from

File: src/libsync/propagateupload.cpp [hunk 3]
```diff
@@ -209,30 +209,45 @@ void PropagateUploadFileQNAM::start()
...
+    if (!_item->_checksumHeader.isEmpty()) {
+        QByteArray checksumType;
+        QByteArray checksum;
+        if (parseChecksumHeader(_item->_checksumHeader, &checksumType, &checksum)
+                && supportedChecksumTypes.contains(checksumType)) {
```
Comment (dragotin) [inline src/libsync/propagateupload.cpp]: I would love to see that split into two lines to make it more clear that the `contains()` relies on having `checksumType` set in the call to `parseChecksumHeader()`.

**Comment by dragotin** [comment]
Apart from my  minor comments: :+1:

File: src/libsync/capabilities.cpp (hunk 1 above)
Comment (ckamm) [inline src/libsync/capabilities.cpp]: Yes. This is mainly intended for testing and will definitely need to be revised as the server API comes into existence and as we figure out how exactly we want to use the checksums.

**Comment by dragotin** [comment]
:+1:

**Pull request merged**